File: iec61131lang/src/main/java/edu/kit/iti/formal/automation/st/LookupList.java
package edu.kit.iti.formal.automation.st;

import java.util.List;
import java.util.Optional;

/**
 * @author Alexander Weigl
 * @version 1 (09.05.18)
 */
public interface LookupList<T extends Identifiable> extends List<T> {
    default Optional<T> get(String name) {
        return stream().findFirst().filter(i -> i.getIdentifier().equals(name));
    }

    default Optional<T> getIgnoreCase(String name) {
        return stream().findFirst().filter(i -> i.getIdentifier().equalsIgnoreCase(name));
    }
}

File: iec61131lang/src/main/kotlin/edu/kit/iti/formal/automation/st/LookupList.kt
package edu.kit.iti.formal.automation.st

/**
 * @author Alexander Weigl
 * @version 1 (09.05.18)
 */
class LookupList<T>(private val impl: ArrayList<T> = arrayListOf()) : MutableCollection<T> by impl
        where T : Cloneable, T : Identifiable {

    constructor(sz: Int) : this() {
        impl.ensureCapacity(sz)
    }

    constructor(seq: Collection<T>) : this(seq.size) {
        addAll(seq)
    }

    operator fun get(name: String): T? = this[{ obj -> obj == name }]
    operator fun get(x: (String) -> Boolean): T? = find { x(it.name) }
    fun getIgnoreCase(name: String) =
            this[{ obj -> obj.equals(name, ignoreCase = true) }]

    operator fun contains(name: String) = this[name] != null
    fun remove(element: String): Boolean {
        val obj = this[element]
        if (obj != null) return remove(obj)
        return false
    }

    fun replace(name: String, variable: T): Boolean {
        remove(name)
        return add(variable)
    }

    fun clone(): LookupList<T> {
        val new = LookupList<T>(this.size)
        forEach { new += it.clone() as T }
        return new
    }

    override fun toString() = impl.toString()
}
